File: cpp/src/material.cpp
#include "all.h"
#include <Eigen/Core>
#include "material.h"

using EV = Eigen::VectorXd ;
using EM = Eigen::MatrixXd ;
using EM3 = Eigen::Matrix3d ;

Material::Material () {}
Material::Material (size_t id, std::string style, std::vector<double> param) {
  this->id = id;
  this->style = style;
  this->param = param;

  this->set_param();
}

void Material::set_init(size_t id, std::string style, std::vector<double> param) {
  this->id = id;
  this->style = style;
  this->param = param;

  this->set_param();
}

void Material::print() {
  std::cout << this->id << ": ";
  std::cout << this->style << ", ";
  for (size_t i = 0 ; i < this->param.size() ; i++) {
    std::cout << this->param.at(i) << " ";
  }
  std::cout << "\n";
}

void Material::set_param() {
  if (this->style == "vs_vp_rho") {
    double vs = this->param.at(0);
    double vp = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = rho*vs*vs;
    this->rlambda = rho*vp*vp - 2.0*this->rmu;
    this->rho = rho;

  } else if (this->style == "nu_vp_rho") {
    double nu = this->param.at(0);
    double vp = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = rho/2.0*vp*vp*(1.0-2.0*nu)/(1.0-nu);
    this->rlambda = rho*nu*vp*vp/(1.0-nu);
    this->rho = rho;

  } else if (this->style == "nu_vs_rho") {
    double nu = this->param.at(0);
    double vs = this->param.at(1);
    double rho = this->param.at(2);

    this->rmu = rho*vs*vs;
    this->rlambda = 2.0*nu/(1.0-2.0*nu) * this->rmu;
    this->rho = rho;

  } else if (this->style == "vs_vp_rho_Q") {
    double vs = this->param.at(0);
    double vp = this->param.at(1);
    double rho = this->param.at(2);
    double Q = this->param.at(3);
    double fmax = this->param.at(4);

    this->rmu = rho*vs*vs;
    this->rlambda = rho*vp*vp - 2.0*this->rmu;
    this->rho = rho; 

    double wmax = 2.0*M_PI*fmax;
    this->beta = 3.0 / (2.0*Q*wmax);
  }
}

// ------------------------------------------------------------------- //
EM Material::mk_d(const size_t dof) {
    EM D(6,6);

    D = EM::Zero(6,6);
    D(0,0) = this->rlambda + 2.0*this->rmu;
    D(0,1) = this->rlambda;
    D(0,2) = this->rlambda;

    D(1,0) = this->rlambda;
    D(1,1) = this->rlambda + 2.0*this->rmu;
    D(1,2) = this->rlambda;

    D(2,0) = this->rlambda;
    D(2,1) = this->rlambda;
    D(2,2) = this->rlambda + 2.0*this->rmu;

    D(3,3) = this->rmu;
    D(4,4) = this->rmu;
    D(5,5) = this->rmu;

    return D;
  }

// ------------------------------------------------------------------- //
EM Material::mk_visco(const size_t dof) {
    EM D(6,6);
    D = EM::Zero(6,6);

    if (this->style.find("Q") != std::string::npos) {
      double d0 =  4.0/3.0*this->rmu;
      double d1 = -2.0/3.0*this->rmu;

      D(0,0) = d0;
      D(0,1) = d1;
      D(0,2) = d1;

      D(1,0) = d1;
      D(1,1) = d0;
      D(1,2) = d1;

      D(2,0) = d1;
      D(2,1) = d1;
      D(2,2) = d0;

      D(3,3) = this->rmu;
      D(4,4) = this->rmu;
      D(5,5) = this->rmu;

      D *= this->beta;
    }

    return D;
  }

// ------------------------------------------------------------------- //
EM3 Material::mk_imp(const size_t dof) {
    EM3 imp;
    double vs = sqrt(this->rmu/this->rho);
    double vp = sqrt((this->rlambda + 2.0*this->rmu)/this->rho);

    imp = EM::Zero(3,3);
    imp(0,0) = this->rho * vp;
    imp(1,1) = this->rho * vs;
    imp(2,2) = this->rho * vs;

    return imp;
  }

File: python/src/material.py
import numpy as np

class Material:
    def __init__(self,id,style,param):
        self.id = id
        self.style = style

        self.set_param(param)

    def print(self):
        print(self.id,":",self.style,self.param)

    def set_param(self,param):
        self.param = param

        if self.style == "vs_vp_rho":
            vs,vp,rho = param
            self.rmu = rho*vs*vs
            self.rlambda = rho*vp*vp - 2.*self.rmu
            self.rho = rho

        elif self.style == "nu_vp_rho":
            nu,vp,rho = param
            self.rmu = rho/2*vp**2*(1-2*nu)/(1-nu)
            self.rlambda = rho*nu*vp**2/(1-nu)
            self.rho = rho

        elif self.style == "nu_vs_rho":
            nu,vs,rho = param
            self.rmu = rho*vs**2
            self.rlambda = 2*nu/(1-2*nu) * self.rmu
            self.rho = rho

        elif self.style == "slip_joint_node_normal":
            self.rho = 0.0
            n0,n1,amp = param

            norm = np.sqrt(n0*n0+n1*n1)
            n0,n1 = n0/norm,n1/norm

            self.R = np.zeros([2,2], dtype=np.float64)
            self.R[0,0],self.R[0,1] =  n0, n1
            self.R[1,0],self.R[1,1] = -n1, n0

    # ---------------------------------------------------------
    def mk_d(self,dof):
        D = np.zeros([6,6],dtype=np.float64)
        D[0,0],D[0,1],D[0,2] = self.rlambda + 2*self.rmu, self.rlambda, self.rlambda
        D[1,0],D[1,1],D[1,2] = self.rlambda, self.rlambda + 2*self.rmu, self.rlambda
        D[2,0],D[2,1],D[2,2] = self.rlambda, self.rlambda, self.rlambda + 2*self.rmu
        D[3,3] = self.rmu
        D[4,4] = self.rmu
        D[5,5] = self.rmu

        return D

    # ---------------------------------------------------------
    def mk_visco(self,dof):
        mu = 0.001 # [Pa s]

        D = np.zeros([6,6],dtype=np.float64)
        D[3,3] = mu
        D[4,4] = mu
        D[5,5] = mu

        return D

    # ---------------------------------------------------------
    def mk_imp(self,dof):
        vs = np.sqrt(self.rmu/self.rho)
        vp = np.sqrt((self.rlambda +2*self.rmu)/self.rho)

        imp = np.diag([self.rho*vp,self.rho*vs,self.rho*vs])

        return imp
